src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,410 @@
+            {
+                BoundExpression savedRight = EvaluateSideEffectingArgumentToTemp(VisitExpression(right), outerEffects, ref initialEffectsAndTemps.temps);
+                rightHasValue = MakeNullableHasValue(savedRight.Syntax, savedRight); // no need for local for right.HasValue since used once
+                rightValue = MakeValueOrDefaultTemp(savedRight, initialEffectsAndTemps.temps, innerEffects);
+                inRightLiteral = false;
+            }
+            else
+            {
+                rightHasValue = MakeBooleanConstant(right.Syntax, true);
+                rightValue = SaveTupleIfNotLiteral(coreRight, ref inRightLiteral, initialEffectsAndTemps, isRight: true);
+            }
+
+            BoundExpression logicalExpression = RewriteNonNullableNestedTupleOperators(operators, leftValue, rightValue, boolType, initialEffectsAndTemps.temps, innerEffects, initialEffectsAndTemps, operatorKind, inLeftLiteral, inRightLiteral);
+
+            BoundExpression innerSequence = _factory.Sequence(ImmutableArray<LocalSymbol>.Empty, innerEffects.ToImmutableAndFree(), logicalExpression);
+
+            if (!isLeftNullable && !isRightNullable)
+            {
+                return innerSequence;
+            }
+
+            // outer sequence:
+            //      leftHasValue == rightHasValue
+            //          ? leftHasValue ? ... inner sequence ... : true/false
+            //          : false/true
+            bool boolValue = operatorKind == BinaryOperatorKind.Equal; // true/false
+            BoundExpression outerSequence =
+                _factory.Sequence(ImmutableArray<LocalSymbol>.Empty, outerEffects.ToImmutableAndFree(),
+                    _factory.Conditional(
+                        _factory.Binary(BinaryOperatorKind.Equal, boolType, leftHasValue, rightHasValue),
+                        _factory.Conditional(leftHasValue, innerSequence, MakeBooleanConstant(right.Syntax, boolValue), boolType),
+                        MakeBooleanConstant(right.Syntax, !boolValue),
+                        boolType));
+
+            return outerSequence;
+        }
+
+        private BoundExpression WithoutImplicitNullableConversions(BoundExpression expr)
+        {
+            while (true)
+            {
+                if (expr.Kind != BoundKind.Conversion)
+                {
+                    return expr;
+                }
+
+                var conversion = (BoundConversion)expr;
+                if (!conversion.Conversion.IsImplicit || !conversion.Conversion.IsNullable)
+                {
+                    return expr;
+                }
+
+                expr = conversion.Operand;
+            }
+        }
+
+        private BoundExpression MakeTemp(BoundExpression expression, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> effects)
+        {
+            BoundLocal temp = _factory.StoreToTemp(VisitExpression(expression), out BoundAssignmentOperator assignmentToTemp);
```

Comment:
It looks like the callers below do not require `expression` to be lowered.

Reply:
Good catch. Fixed

---
In reply to: [168941060](https://github.com/dotnet/roslyn/pull/23356#discussion_r168941060) [](ancestors = 168941060)